src/UglyToad.PdfPig/Writer/PdfStreamWriter.cs

```diff
@@ -0,0 +1,166 @@
+            Stream.WriteByte(210);
+            Stream.WriteNewLine();
+
+            var offsets = new Dictionary<IndirectReference, long>();
+            ObjectToken catalogToken = null;
+            foreach(var pair in tokenReferences)
+            {
+                var referenceToken = pair.Key;
+                var token = pair.Value;
+                var offset = Stream.Position;
+                var obj = new ObjectToken(offset, referenceToken.Data, token);
+
+                TokenWriter.WriteToken(obj, Stream);
+
+                offsets.Add(referenceToken.Data, offset);
+
+                if (catalogToken == null && referenceToken == catalogReference)
+                {
+                    catalogToken = obj;
+                }
+            }
+
+            if (catalogToken == null)
+            {
+                throw new Exception("Catalog object wasn't found");
+            }
+
+            // TODO: Support document information
+            TokenWriter.WriteCrossReferenceTable(offsets, catalogToken, Stream, null);
+        }
+
+        public IndirectReferenceToken WriteToken(IToken token, int? reservedNumber = null)
+        {
+            // if you can't consider deduplicating the token. 
+            // It must be because it's referenced by his child element, so you must have reserved a number before hand
+            // Example /Pages Obj
+            var canBeDuplicated = !reservedNumber.HasValue;
+            if (!canBeDuplicated)
+            {
+                if (!reservedNumbers.Remove(reservedNumber.Value))
+                {
+                    throw new InvalidOperationException("You can't reuse a reserved number");
+                }
+
+                // When we end up writing this token, all of his child would already have been added and checked for duplicate
+                return AddToken(token, reservedNumber.Value);
+            }
+
+            var reference = FindToken(token);
+            if (reference == null)
+            {
+                return AddToken(token, CurrentNumber++);
+            }
+
+            return reference;
+        }
+        
+        public int ReserveNumber()
+        {
+            var reserved = CurrentNumber;
+            reservedNumbers.Add(reserved);
+            CurrentNumber++;
+            return reserved;
+        }
+
+        public IndirectReferenceToken ReserveNumberToken()
+        { 
+            return new IndirectReferenceToken(new IndirectReference(ReserveNumber(), 0));
+        }
+        
+        public byte[] ToArray()
+        {
+            if (!Stream.CanSeek)
+                throw new NotSupportedException($"{Stream.GetType()} can't seek");
+            
+            var currentPosition = Stream.Position;
+            Stream.Seek(0, SeekOrigin.Begin);
+
+            var bytes = new byte[Stream.Length];
+
+            // Should we slice the reading into smaller chunks?
+            if (Stream.Read(bytes, 0, bytes.Length) != bytes.Length)
```

Comment:
One alternative is just to use an intermediary memory stream and let the framework worry about the buffer handling:

```
using (var memory = new MemoryStream())
{
   stream.CopyTo(memory)l
   return memory.ToArray()
}
```

Reply:
I highly against this.  When we execute this method. The result would be at least 3 buffer of the same size
1. PdfStreamWriter stream
2. MemoryStream memory
3. The copy that we return, when we do `memory.ToArray`

That's without counting the list of token that we haven't already flushed... For simple document, this would not matter much, but for big document, the memory footprint would be huge

Reply:
That's fair enough, I'm happy with it as-is.